### PR TITLE
Auth set session manually beta

### LINKED ISSUE #399: currentSession() returns no current user?
I'm trying to test with the following code in a component that is wrapped by the withAuthenticator() HOC. Federated login works. Once logged in, I get the "No current user" error when running the following code, but the user object is getting populated. 

Am I not understanding how the library works? Shouldn't the currentSession return JWT, acesstoken, id?

Ultimately, I want to be able to get email from Facebook as identity provider. 

```
componentDidMount = () => {
    window.LOG_LEVEL = 'DEBUG'
    console.log("About to mount")
    Auth.currentSession().then(function(session) {
      console.log(JSON.stringify(session))
    }, function(err) {
      console.log(err)
    })
    Auth.currentAuthenticatedUser().then(function(user) {
      console.log("USER: " + JSON.stringify(user))
    })
  }
```

Output:
```
No current user
USER: {"Name":"x x"}
```

### LINKED ISSUE #825: Possible to set current credentials manually?
Is it possible to set the credentials for the current user manually in an easy way?

I have a custom login using the authorize and token endpoints (custom login interface), and wish to load the credentials in the Auth library to continue using it in the rest of my app.

The Auth lib has support for parsing a request from the hosted UI containing code and state parameters. It then makes a post to the TOKEN endpoint and loads the credentials. It doesn't seem to have support for PKCE sadly and it doesn't check the STATE parameter either I think to make sure it's correct? see getCodeQueryParameter in CognitoAuth.js for the parsing.

So I am wondering if it's possible to supply the JWT token to Auth for it to load the user etc.
I basically want to call CognitoAuth.onSuccessExchangeForToken

Is there an easy way to get access to the CognitoAuth instance in the Auth component (import { Auth } from 'aws-amplify';) so I can call its methods? One idea could be to expose a method for calling the /authorize endpoint yourself so you can initiate an authorization flow. As well as adding PKCE support for the flow. The rest of the functionality seems to be there already

### PR DESCRIPTION
**Pull request by powerful23**
*Issue #, if available:*
fixes #825 
fixes #399 
*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/auth/src/Auth.ts [hunk 1]
```diff
@@ -1073,71 +1257,68 @@ export default class AuthClass {
...
+     * @async
+     * @return {Object }- current User's information
      */
-    public verifyUserAttributeSubmit(user: CognitoUser | any, attr: string, code: string): Promise<string> {
-        if (!code) { return Promise.reject('Code cannot be empty'); }
+    public async currentUserInfo() {
+        const sessionSource = this._getSessionSource();
+         if (!sessionSource || sessionSource === AWSCognitoProvider.NAME) {
+            const user = await this.currentUserPoolUser()
+                .catch(err => logger.debug(err));
```
Comment (manueliglesias) [inline packages/auth/src/Auth.ts]: shouldn't this `catch(...)` be an actual `try {} catch(err) {}` block?

File: packages/auth/src/Auth.ts [hunk 2]
```diff
@@ -1187,15 +1368,24 @@ export default class AuthClass {
             logger.debug('failed to clear cached items');
         }
 
-        if (this.userPool) { 
+        await this._storageSync;
+        const sessionSource = this._getSessionSource();
+        this._removeSessionSource();
+        if (!sessionSource || sessionSource === AWSCognitoProvider.NAME) {
+            if (!this.userPool) {
+                // throw('No userPool');
```
Comment (manueliglesias) [inline packages/auth/src/Auth.ts]: Is this intentional? should it be removed/uncommented?

File: packages/auth/src/Auth.ts [hunk 3]
```diff
@@ -1341,25 +1407,69 @@ export default class AuthClass {
      * @param {String} user - user info
      */
     public async federatedSignIn(
-        provider: 'google'|'facebook'|'amazon'|'developer'|string, 
+        provider: 'Google'|'Facebook'|'Amazon'|'Developer'|string,
```
Comment (manueliglesias) [inline packages/auth/src/Auth.ts]: Should these providers be exported as an enum?

File: packages/auth/src/Auth.ts [hunk 4]
```diff
@@ -1341,25 +1407,69 @@ export default class AuthClass {
...
-        // Because Credentials.set would update the user info with identity id
-        // So we need to retrieve the user again.
-        const credentials = await Credentials.set({ provider, token, identity_id, user, expires_at }, 'federation');
-        const currentUser = await this.currentAuthenticatedUser();
-        dispatchAuthEvent('signIn', currentUser);
+        let authProvider = null;
+        let domain = null;
+
+        // for backward compatiblity
+        switch(provider) {
```
Comment (manueliglesias) [inline packages/auth/src/Auth.ts]: Having `provider` type as the enum from the previous comment would make typescript help us make sure we are exhaustive with the values. 

See [Exhaustiveness checking]

File: packages/auth/src/Auth.ts [hunk 5]
```diff
@@ -1408,4 +1518,58 @@ export default class AuthClass {
...
+        if (!providerClass) {
+            throw new Error(`No AuthProvider class for the provider, ${provider}`);
+        }
+
+        try {
+            const result = await providerClass.setSession(params);
+            this._addSessionSource(provider);
+            return result;
+        } catch (e) {
+            throw e;
```
Comment (manueliglesias) [inline packages/auth/src/Auth.ts]: Should we be logging this error?

File: packages/auth/src/providers/AWSCognitoProvider.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        
+        return {
+            session: userSession,
+            user,
+            credentials
+        };
+    }
+
+    public async getSession(): Promise<any> {
+        logger.debug('tobe implemented');
```
Comment (manueliglesias) [inline packages/auth/src/providers/AWSCognitoProvider.ts]: Are these going to be implemented as part of this PR?

File: packages/auth/src/providers/BaseProvider.ts [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+    public async setSession(params: ExternalSession): Promise<SetSessionResult> {
+        const { _keyPrefix } = this._config;
+        const { 
+            username, 
+            attributes, 
+            tokens, 
+            expires_at,
+            provider,
+            federatedWithIDP={}
```
Comment (manueliglesias) [inline packages/auth/src/providers/BaseProvider.ts]: You can provide nested default values too

```ts
const {
    federatedWithIDP: {
        domain = this._credentialsDomain,
        token = this._credentialsTokenSource || BaseProvider.ID_TOKEN
    }
 } = params;

```

File: packages/auth/src/Auth.ts [hunk 8]
```diff
@@ -447,6 +497,62 @@ export default class AuthClass {
...
 
+    /**
+     * Send MFA code to confirm sign in
+     * @param {Object} user - The CognitoUser object
+     * @param {String} code - The confirmation code
+     */
+    public confirmSignIn(
+        user: CognitoUser | any, 
+        code: string, 
+        mfaType?: 'SMS_MFA'|'SOFTWARE_TOKEN_MFA'|null
```
[the remark this answers is not in the document]
Comment (powerful23) [reply packages/auth/src/Auth.ts]: Good point. Done.

File: packages/auth/src/Auth.ts [hunk 9]
```diff
@@ -447,6 +497,62 @@ export default class AuthClass {
         });
     }
 
+    /**
+     * Send MFA code to confirm sign in
+     * @param {Object} user - The CognitoUser object
+     * @param {String} code - The confirmation code
+     */
+    public confirmSignIn(
+        user: CognitoUser | any,
```
[the remark this answers is not in the document]
Comment (powerful23) [reply packages/auth/src/Auth.ts]: @manueliglesias because of some historical issue. We added some extra properties into the user object returned from the Cognito js sdk when calling `currentAuthenticatedUser()` which makes the user returned from that function is not recognized by typescript.

File: packages/auth/src/Auth.ts (hunk 3 above)
Comment (powerful23) [reply packages/auth/src/Auth.ts]: Done.

File: packages/auth/src/Auth.ts (hunk 5 above)
Comment (powerful23) [reply packages/auth/src/Auth.ts]: Done.

File: packages/auth/src/providers/AWSCognitoProvider.ts (hunk 6 above)
Comment (powerful23) [reply packages/auth/src/providers/AWSCognitoProvider.ts]: Because we are still using the Cognito's sdk. We can implement those functions after we start implement our own Cognito sdk or in another pr. I don't want to make this one too big.

File: packages/auth/src/providers/BaseProvider.ts (hunk 7 above)
Comment (powerful23) [reply packages/auth/src/providers/BaseProvider.ts]: Will it overwrite the whole `federatedWithIDP` object? Because there are some other properties in that object.

File: packages/auth/src/Auth.ts (hunk 4 above)
Comment (powerful23) [reply packages/auth/src/Auth.ts]: The type of `provider` here is `FederatedProvider | string`. Because it could be arbitrary string, I don't think it's possible to apply the Exhaustiveness checking here. What do you think?

File: packages/auth/src/providers/AWSCognitoProvider.ts (hunk 6 above)
Comment (manueliglesias) [reply packages/auth/src/providers/AWSCognitoProvider.ts]: Alright, sounds good

File: packages/auth/src/Auth.ts [hunk 10]
```diff
@@ -447,6 +497,62 @@ export default class AuthClass {
...
+                code, {
+                    onSuccess: async (session) => {
+                        logger.debug(session);
+                        this._addSessionSource(AWSCognitoProvider.NAME);
+                        try {
+                            await Credentials.clear();
+                            const cred = await Credentials.set(session, 'session');
+                            logger.debug('succeed to get cognito credentials', cred);
+                        } catch (e) {
+                            logger.debug('cannot get cognito credentials', e);
```
[the remark this answers is not in the document]
Comment (manueliglesias) [reply packages/auth/src/Auth.ts]: cool

File: packages/auth/src/Auth.ts [hunk 11]
```diff
@@ -447,6 +497,62 @@ export default class AuthClass {
...
+                mfaType);
+        });
+    }
+
+    /**
+     * Send the answer to a custom challenge
+     * @param {CognitoUser} user - The CognitoUser object
+     * @param {String} challengeResponses - The confirmation code
+     */
+    public sendCustomChallengeAnswer(user:CognitoUser | any, challengeResponses: string): Promise<CognitoUser | any> {
```
[the remark this answers is not in the document]
Comment (manueliglesias) [reply packages/auth/src/Auth.ts]: answered above

File: packages/auth/src/Auth.ts (hunk 3 above)
Comment (manueliglesias) [inline packages/auth/src/Auth.ts]: I think this should be an enum,  e.g.

```ts
enum Providers {
    Google = 'google',
    Facebook = 'facebook',
    Amazon = 'amazon',
    Developer = 'developer',
}

const federatedSignIn = (provider: Providers | string) => {
    switch (provider) {
        case '':
        case null:
        case undefined:
            throw new Error('Invalid provider');
        case Providers.Google:
        case Providers.Facebook:
        case Providers.Amazon:
        case Providers.Developer:
            console.log('predefined provider');
            break;
        default:
            console.log('custom provider');
    }
}
```

File: packages/auth/src/Auth.ts (hunk 4 above)
Comment (manueliglesias) [reply packages/auth/src/Auth.ts]: see updated comment

File: packages/auth/src/Auth.ts (hunk 3 above)
Comment (powerful23) [reply packages/auth/src/Auth.ts]: @manueliglesias this should have been changed: https://github.com/aws-amplify/amplify-js/pull/2569/files#diff-1cf5b10c87e919f81d5f1ea503f15074R1412

File: packages/auth/src/Auth.ts [hunk 12]
```diff
@@ -793,20 +863,90 @@ export default class AuthClass {
...
+        if (!username) { return Promise.reject('Username cannot be empty'); }
 
-        const that = this;
+        const user = this.createCognitoUser(username);
         return new Promise((resolve, reject) => {
-            user.sendCustomChallengeAnswer(challengeResponses, this.authCallbacks(user, resolve, reject));
+            user.forgotPassword({
+                onSuccess: () => { 
+                    resolve();
+                    return;
```
[the remark this answers is not in the document]
Comment (powerful23) [reply packages/auth/src/Auth.ts]: I want to leave it there to prevent someone not familiar with `Promise` to do something wrong.

File: packages/auth/src/Auth.ts (hunk 1 above)
Comment (powerful23) [reply packages/auth/src/Auth.ts]: Good catch. Done.

File: packages/auth/src/Auth.ts (hunk 2 above)
Comment (powerful23) [reply packages/auth/src/Auth.ts]: Good catch. We should throw the error when user pool is not configured.

File: packages/auth/src/providers/AWSCognitoProvider.ts (hunk 6 above)
Comment (rubene) [reply packages/auth/src/providers/AWSCognitoProvider.ts]: @powerful23 @manueliglesias Do you have a timeline for the implementation? @mvandenbeuken

**Comment by kurbur** [comment]
@manueliglesias how close is this to being merged into master?

**Comment by kheriox-technologies** [comment]
is it merged to master yet ? We are using @aws-amplify/auth in our react application.

**Comment by kurbur** [comment]
> is it merged to master yet ? We are using @aws-amplify/auth in our react application.

@kheriox-technologies , If you are just looking to RESET auth session, the following will work. If you are looking to specifically set auth credentials, then this will not work. 

Amplify uses 'Credentials' and 'StorageHelper' under the hood. It uses Credentials to generate unauthenticated session tokens/credentials and Storage helper to store them locally and access them later (so that it doesn't have to regenerate those credentials every time).  To reset those credentials, you clear them from Credentials and wipe them from local storage with StorageHelper so that it will be forced to regenerate new tokens/credentials.

`import { Credentials, StorageHelper, } from '@aws-amplify/core';

export default () => new Promise(async (resolve, reject) => {
  try {
    await Credentials.clear();
    const storage = new StorageHelper().getStorage();
    const identityIdKey = 'CognitoIdentityId-' + AmplifyConfig.Auth.identityPoolId;

    let identityId = storage.getItem(identityIdKey);
    console.log('removing cached credentials', {identityId});

    storage.removeItem(identityIdKey);
    resolve()
  } catch (err) {
    reject(err)
  }
})`

BTW, AmplifyConfig is just my own local config file. nothing special. 

For good measure, I also call 'Auth.getCurrentCredentials()' because it checks local storage for credentials and if they are not there, it will regenerate new ones. 

`import {Auth} from 'aws-amplify'

export default () => new Promise((resolve, reject) => {
  Auth.currentUserCredentials({bypassCache: true})
    .then(data => {
      console.log("got current credentials", data)
      resolve(data)
    })
    .catch(err => {
      reject(err)
    })
})`

So TLDR, this will remove cached credentials and refresh them. I hope this helps.
